src/data/nat/factorization.lean

```diff
@@ -408,18 +410,32 @@ def rec_on_prime_pow {P : ℕ → Sort*} (h0 : P 0) (h1 : P 1)
 `P b` to `P (a * b)` when `a, b` are positive coprime, we can define `P` for all natural numbers. -/
 @[elab_as_eliminator]
 def rec_on_pos_prime_pos_coprime {P : ℕ → Sort*} (hp : ∀ p n : ℕ, prime p → 0 < n → P (p ^ n))
-  (h0 : P 0) (h1 : P 1) (h : ∀ a b, 0 < a → 0 < b → coprime a b → P a → P b → P (a * b)) :
+  (h0 : P 0) (h1 : P 1) (h : ∀ a b, 1 < a → 1 < b → coprime a b → P a → P b → P (a * b)) :
   ∀ a, P a :=
-rec_on_prime_pow h0 h1 $ λ a p n hp' hpa ha,
-  (h (p ^ n) a (pow_pos hp'.pos _) (nat.pos_of_ne_zero (λ t, by simpa [t] using hpa))
-  (prime.coprime_pow_of_not_dvd hp' hpa).symm
-  (if h : n = 0 then eq.rec h1 h.symm else hp p n hp' $ nat.pos_of_ne_zero h) ha)
+rec_on_prime_pow h0 h1 $
+begin
+  intros a p n hp' hpa hn hPa,
+  by_cases ha1 : a = 1,
+  { rw [ha1, mul_one],
+    apply hp p n hp' hn, },
+  apply h (p^n) a
+    (lt_of_lt_of_le (hp'.one_lt) (le_self_pow (le_of_lt (prime.one_lt hp')) (succ_le_iff.mpr hn)))
+    _ _ (hp _ _ hp' hn) hPa,
+  { by_contra,
+    simp only [not_lt] at h,
+    interval_cases a,
+    simp only [dvd_zero, not_true] at hpa,
+    assumption,
+    apply ha1,
+    dec_trivial, },
+  simpa [hn, prime.coprime_iff_not_dvd hp'],
```

Comment:
A bit of golfing. My main remark here is to be sure to name the hypotheses that tactics like `by_contra` introduce, for extra robustness.
```suggestion
  { refine lt_of_not_ge (λ (h : a ≤ 1), _),
    interval_cases a,
    { simpa only [dvd_zero, not_true] using hpa },
    { contradiction } },
  simpa [hn, prime.coprime_iff_not_dvd hp'],
```